kyo-os-lib/shared/src/main/scala/kyo/os.scala

```diff
@@ -0,0 +1,213 @@
+package kyo
+
+import java.io.*
+import java.lang.Process as JProcess
+import java.lang.ProcessBuilder.Redirect
+import java.nio.charset.Charset
+import java.nio.charset.StandardCharsets
+import java.nio.file.Path
+import java.util.concurrent.TimeUnit
+import kyo.*
+import scala.jdk.CollectionConverters.*
+
+sealed trait Command:
+    self =>
+    import Command.*
+
+    /** Spawns a new process executing this command
+      */
+    def spawn: Process < (IOs)
+
+    /** Spawns a new process and returns its output as a String
+      */
+    def string: String < (IOs) =
+        stream.map { inputStream =>
+            new String(inputStream.readAllBytes())
+        }
+
+    /** Spawns a new process and returns its output as a Stream
+      */
+    def stream =
+        spawn.map(_.stdout)
+
+    def exitValue =
+        spawn.map(_.exitValue)
+
+    def waitFor =
+        spawn.map(_.waitFor)
+
+    def waitFor(timeout: Long, unit: TimeUnit) =
+        spawn.map(_.waitFor(timeout, unit))
+
+    def pipe(that: Command): Command =
+        Command.Piped(self.flatten ++ that.flatten).simplify
+
+    def >>(that: Command): Command =
+        self.pipe(that)
+
+    def flatten: List[Command.Simple]
+
+    def +(that: Command): Command =
+        val cs = self.flatten ++ that.flatten
+        if cs.length > 1 then Command.Piped(cs) else cs.head
+
+    def map(f: Command.Simple => Command): Command
+
+    def simplify: Command =
+        self match
+            case Command.Piped(cs) if cs.length == 1 => cs.head
+            case c                                   => c
+
+    /** Modifies the current working directory of the command
+      */
+    def cwd(newCwd: Path): Command
+
+    /** Modifies the environment variables of the command
+      */
+    def env(newEnv: Map[String, String]): Command
+
+    /** Modifies the `stdin` of the command
+      */
+    def stdin(newStdin: ProcessInput): Command
+
+    /** Modifies the `stdout` of the command
+      */
+    def stdout(newStdout: ProcessOutput): Command
+
+    /** Modifies the `stderr` of the command
+      */
+    def stderr(newStderr: ProcessOutput): Command
+
+    /** Changes `redirectErrorStream` property
+      */
+    def redirectErrorStream(value: Boolean): Command
+
+end Command
+
+object Command:
+
+    def apply(commands: String*): Command.Simple =
+        Command.Simple(commands.toList)
+
+    case class Simple(
+        command: List[String],
+        cwd: Option[Path] = None,
+        env: Map[String, String] = Map.empty,
+        stdin: ProcessInput = ProcessInput.Inherit,
+        stdout: ProcessOutput = ProcessOutput.Pipe,
+        stderr: ProcessOutput = ProcessOutput.Pipe,
+        redirectErrorStream: Boolean = false
+    ) extends Command:
+        self =>
+        override def spawn: Process < (IOs) =
+            for
+                process <- IOs {
+                    val builder = new ProcessBuilder(command*)
```

Comment:
It doesn't need to be in the scope of this change but something that would be useful to have is a way to fork a `Function0` into a separate JVM process. We'd need to find the current java binary, setup the classpath, and pass the serialized function to be executed.